src/libs/antares/study/include/antares/study/area/reserveParticipationContainer.h

```diff
@@ -0,0 +1,206 @@
+** it under the terms of the Mozilla Public Licence 2.0 as published by
+** the Mozilla Foundation, either version 2 of the License, or
+** (at your option) any later version.
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+
+#include <ranges>
+
+#include <antares/study/area/reserveParticipation.h>
+
+template<typename T>
+concept DerivedFromReserveParticipationBase = std::is_base_of_v<ReserveParticipationBase, T>;
+
+/// @brief Container for reserve participation generic for each type of cluster (thermal, Short
+/// term, Long term)
+template<DerivedFromReserveParticipationBase T>
+class ReserveParticipationContainerBase
+{
+    struct ReserveParticipationWithName
+    {
+        T* reserveParticipation;
+        ReserveName reserveName;
+    };
+
+protected:
+    std::map<ReserveName, T> reservesParticipations;
+    std::vector<std::vector<ReserveParticipationWithName>> reserveParticipationsSymmetries;
+
+public:
+    /// @brief Add a reserve participation to the container for a given reserve name
+    /// @param reserveName name of the reserve for which the participation is added
+    /// @param reserveParticipation reserve participation to add
+    void addReserveParticipation(ReserveName reserveName, const T& reserveParticipation)
+    {
+        reservesParticipations.emplace(reserveName, reserveParticipation);
+    }
+
+    /// @brief Add a reserve participation symmetry to the container
+    /// @param names names of the reserves for which the participation is symmetrical
+    void addReserveParticipationSymmetry(std::set<ReserveName> names)
+    {
+        auto symmetryRes = std::vector<ReserveParticipationWithName>();
+        for (auto name: names)
+        {
+            if (names.size() != 2)
+            {
+                throw std::out_of_range(
+                  "Must have two distinct reserves to participate to a symmetry");
+            }
+            if (reservesParticipations.contains(name))
+            {
+                symmetryRes.push_back(
+                  ReserveParticipationWithName{&reservesParticipations.at(name), name});
+            }
+            else
+            {
+                throw std::out_of_range("This entity is not participating to reserve " + name);
+            }
+        }
+        for (auto existingSymmetry: reserveParticipationsSymmetries)
+        {
+            int sameNamesCount = 0;
+            for (auto resSym: existingSymmetry)
+            {
+                for (auto name: names)
+                {
+                    if (name == (std::string)resSym.reserveName)
```

Comment:
You wrote this using ReserveName = std::string; i suggest you either delete the using or the ugly cast